packages/engine/Source/Scene/VoxelTraversal.js

```diff
@@ -417,61 +423,36 @@ function requestData(that, keyframeNode) {
     return;
   }
 
-  const provider = that._primitive._provider;
+  const primitive = that._primitive;
+  const provider = primitive.provider;
 
   function postRequestSuccess(result) {
     that._simultaneousRequestCount--;
-    const length = provider.types.length;
-
-    if (!defined(result)) {
-      keyframeNode.state = KeyframeNode.LoadState.UNAVAILABLE;
-    } else if (result === KeyframeNode.LoadState.FAILED) {
-      keyframeNode.state = KeyframeNode.LoadState.FAILED;
-    } else if (!Array.isArray(result) || result.length !== length) {
-      // TODO should this throw runtime error?
-      keyframeNode.state = KeyframeNode.LoadState.FAILED;
-    } else {
-      const megatextures = that.megatextures;
-      for (let i = 0; i < length; i++) {
-        const { voxelCountPerTile, channelCount } = megatextures[i];
-        const { x, y, z } = voxelCountPerTile;
-        const tileVoxelCount = x * y * z;
-
-        const data = result[i];
-        const expectedLength = tileVoxelCount * channelCount;
-        if (data.length === expectedLength) {
-          keyframeNode.metadata[i] = data;
-          // State is received only when all metadata requests have been received
-          keyframeNode.state = KeyframeNode.LoadState.RECEIVED;
-        } else {
-          keyframeNode.state = KeyframeNode.LoadState.FAILED;
-          break;
-        }
-      }
-    }
+    keyframeNode.content = result;
+    keyframeNode.state = defined(result)
+      ? KeyframeNode.LoadState.PROCESSING
+      : KeyframeNode.LoadState.UNAVAILABLE;
   }
 
-  function postRequestFailure() {
+  function postRequestFailure(error) {
     that._simultaneousRequestCount--;
     keyframeNode.state = KeyframeNode.LoadState.FAILED;
   }
 
   const { keyframe, spatialNode } = keyframeNode;
-  const promise = provider.requestData({
+  const requestParameters = {
     tileLevel: spatialNode.level,
     tileX: spatialNode.x,
     tileY: spatialNode.y,
     tileZ: spatialNode.z,
     keyframe: keyframe,
-  });
-
-  if (defined(promise)) {
-    that._simultaneousRequestCount++;
-    keyframeNode.state = KeyframeNode.LoadState.RECEIVING;
-    promise.then(postRequestSuccess).catch(postRequestFailure);
-  } else {
-    keyframeNode.state = KeyframeNode.LoadState.FAILED;
-  }
+  };
+  that._simultaneousRequestCount++;
+  keyframeNode.state = KeyframeNode.LoadState.RECEIVING;
+  provider
+    .requestData(requestParameters)
+    .then(postRequestSuccess)
+    .catch(postRequestFailure);
```

Comment:
I think the (right now hypothetical) case where `provider.requestData` returns `undefined` can be accounted for by setting the state back to `UNLOADED`. But please let me know if I missed something.

```suggestion
  const promise = provider
    .requestData(requestParameters);
  if (!defined(promise)) {
    keyframeNode.state = KeyframeNode.LoadState.UNLOADED;
    return;
  }
  
  promise.then(postRequestSuccess)
    .catch(postRequestFailure);
```